src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/Event.scala

```diff
@@ -0,0 +1,219 @@
+                 derived_tstamp:           Option[Instant],
+                 event_vendor:             Option[String],
+                 event_name:               Option[String],
+                 event_format:             Option[String],
+                 event_version:            Option[String],
+                 event_fingerprint:        Option[String],
+                 true_tstamp:              Option[Instant]) {
+
+  def inventory: Set[Data.ShreddedType] = {
+    val unstructEvent = unstruct_event
+      .data
+      .toSet
+      .map((ue: SelfDescribingData[Json]) => Data.ShreddedType(Data.UnstructEvent, ue.schema))
+
+    val derivedContexts = derived_contexts
+      .data
+      .toSet
+      .map((ctx: SelfDescribingData[Json]) => Data.ShreddedType(Data.Contexts(Data.DerivedContexts), ctx.schema))
+
+    val customContexts = contexts
+      .data
+      .toSet
+      .map((ctx: SelfDescribingData[Json]) => Data.ShreddedType(Data.Contexts(Data.CustomContexts), ctx.schema))
+
+    customContexts ++ derivedContexts ++ unstructEvent
+  }
+}
+
+
+object Event {
+  case class Lossy(event: Event) extends AnyVal
+  case class Lossless(event: Event) extends AnyVal
+
+  implicit val jsonEncoder: Encoder[Event] = deriveEncoder[Event]
+  implicit val losslessEncoder: Encoder[Lossless] = Encoder.instance { lossless: Lossless => lossless.event.asJson }
+  implicit val lossyEncoder: Encoder[Lossy] = Encoder.instance { lossy: Lossy =>
+    val event = lossy.event
+    var json = event.asJson.asObject.get.toVector
+    json = json.patch(json.indexWhere(_._1 == "contexts"), event.contexts.data.groupBy(x => (x.schema.vendor, x.schema.name, x.schema.format, x.schema.version.model)).map {
+      case ((vendor, name, format, model), data) =>
+        (transformSchema("contexts", vendor, name, model), data.map{selfdesc => selfdesc.data}.asJson)
+    }.toVector, 1)
+    json = json.patch(json.indexWhere(_._1 == "unstruct_event"), event.unstruct_event.data.map {
+      case SelfDescribingData(schema, data) =>
+        (transformSchema("unstruct_event", schema.vendor, schema.name, schema.version.model), data)
+    }.toVector, 1)
+    json = json.patch(json.indexWhere(_._1 == "derived_contexts"), event.derived_contexts.data.groupBy(x => (x.schema.vendor, x.schema.name, x.schema.format, x.schema.version.model)).map {
+      case ((vendor, name, format, model), data) =>
+        (transformSchema("contexts", vendor, name, model), data.map{selfdesc => selfdesc.data}.asJson)
+    }.toVector, 1)
+    /*event.unstruct_event.data.foreach{
+      case SelfDescribingData(schema, data) =>
+        json = json.add(transformSchema("unstruct_event", schema.vendor, schema.name, schema.version.model), data)
+    }*/
+    JsonObject(json: _*).asJson
+  }
+  private val parser: Parser[Event] = Parser.deriveFor[Event].get
+
+  def parse(line: String): RowDecodeResult[Event] =
+    parser.parse(line)
+
+  def asJson(event: Event, lossy: Boolean = false): Json =
```

Comment:
Default arguments are code smell.

Reply:
Why is it `asJson` in companion object instead of `asJson` method of `Event` itself? We're passing `Event` as first argument anyway.